reco_utils/recommender/rlrmc/RLRMCalgorithm.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import random
+import numpy as np
+import pandas as pd
+import warnings
+import sys
+# sys.path.append("./pymanopt/")
+# The Pymanopt code has been used and was downloaded via: pip install pymanopt  
+# Online code of Pymanopt: https://github.com/pymanopt/pymanopt
+# Online license link: https://github.com/pymanopt/pymanopt/blob/master/LICENSE
+# Pymanopt is licensed under the BSD 3-Clause "New" or "Revised" License.
+
+from pymanopt import Problem
+from pymanopt.solvers import ConjugateGradientMS # Modified Conjugate Gradient 
+from pymanopt.solvers.linesearch import LineSearchBackTracking
+from pymanopt.manifolds import Stiefel, PositiveDefinite, Product
+from math import sqrt
+from scipy.sparse import csr_matrix
+from scipy.sparse.linalg import svds
+import time 
+from numba import jit, njit, prange
+
+class RLRMCalgorithm(object):
+    """
+    classdocs
+    """
+
+    def __init__(
+        self,
+        rank,
+        C,
+        model_param,
+        initialize_flag = 'random',
+        max_time = 1000,
+        maxiter = 100,
+        seed = 42,
+    ):
+        """
+        Constructor
+        """
+        self.model_param = model_param
+        self.train_mean = model_param.get('train_mean')
+        self.initialize_flag = initialize_flag
+        self.rank = rank
+        self.C = C
+        self.max_time = max_time
+        self.maxiter = maxiter
+
+    def _init_train(self,entries_train_csr):
+        print("Hyper-parameters of the algorithm")
+        print("Rank: %i, Regularization parameter: %e" % (self.rank, self.C))
+        # Initialization # starting point on the manifold
+        initialize_dict = {'random': 0, 'svd': 1}
+        if initialize_dict.get(self.initialize_flag)==0:#rndom
+            W0=None
+        elif initialize_dict.get(self.initialize_flag)==1:#svd
+            U0, B0, V0 = svds(entries_train_csr, k=self.rank)
+            W0 = [U0, V0.T, np.diag(B0)]
+        else: # default option when given incorrect option
+            print("Initialization flag not recognized. Setting it to random (default).")
+            W0=None
+        return W0
+
+    def fit(self,RLRMCdata,verbosity = 0,compute_iter_rmse = False):
```

Comment:
data indexing and mean centering are not the part of the algorithm and hence they are kept as a part of data structure creation. Computationally, it seems more reasonable since any user may want to run RLRMC multiple number of time for a given dataset with different hyperparameter values (e.g., rank, regularization parameter C). If we keep data indexing and mean centering as a part of algorithm, then this may be an additional but unnecessary overhead for the algorithm.